### PR TITLE
PolkadotJS hardware wallets support

### PR DESCRIPTION
**Pull request by akru**
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/8Wx7nxtq8fKrsuEz8) has been filled out for this milestone. 
- [x] This pull request is done by the same account, which is responsible for the pull request of the accepted application. 
- [x] In the case of acceptance, the payment will be transferred to the initial BTC payment address. 
- [x] The delivery is according to the [milestone delivery guidelines](https://github.com/w3f/General-Grants-Program/blob/unify-guidelines/grants/milestone-deliverables-guidelines.md). 

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/39

### CODE REVIEW

**Comment by semuelle** [comment]
Thanks for the delivery, @akru. We will review both milestones as soon as possible.

**Comment by SotaWatanabe** [comment]
Hello guys. I am very sure that you guys are very busy right now. But would it be possible to let us know the status of this review? Thank you very much in advance.

**Comment by semuelle** [comment]
Hi @SotaWatanabe, sorry for the radio silence. It's the very next project on our agenda. Someone will get in touch with you next week.

**Comment by semuelle** [comment]
Hi @SotaWatanabe, the links to the documentation are 404ing; I assume they were moved. Could you update the document?

**Comment by akru** [comment]
Hello @semuelle, sorry for the inconvenience, links fixed.

**Comment by semuelle** [comment]
Hi @akru, thanks for the fix. No worries. 

Just FYI, the review is in progress ([[1]](https://github.com/w3f/Grant-Milestone-Delivery/blob/d63fb6d389ef60e582b99e698e542aaaa506016d/evaluations/stake-polkadotjs-ecdsa-hardware_1_semuelle.md), [[2]](https://github.com/w3f/Grant-Milestone-Delivery/blob/d63fb6d389ef60e582b99e698e542aaaa506016d/evaluations/stake-polkadotjs-ecdsa-hardware_2_semuelle.md)) but will take a couple more days. Sorry for the inconvenience.

**Comment by SotaWatanabe** [comment]
Hello @semuelle I hope you are doing well. I am sure that Web3 Foundation is really busy now since the ecosystem is growing very fast. I don't mean to rush you, but could you share the status of this review since the last reply was on Feb 27th?

**Comment by semuelle** [comment]
Hi @SotaWatanabe & @akru, sincere apologies for the delay on this. 

The good news is, @Lederstrumpf and I just talked about it last Friday and while my last estimate was way off, I expect this to be done _very_ soon.

**Comment by SotaWatanabe** [comment]
Hello @semuelle Thank you very much for your reply. That is good to know. Feel free to ask any technical questions here if you have any :D

**Comment by Lederstrumpf** [comment]
Hi @akru and @SotaWatanabe,

My sincere apologies that this review has been delayed a lot.
I've now gotten hold of a Trezor and am happy to confirm that the signing process works.

I ran into the following two issues though:
1. When aborting a transaction on the device, the `Sign Transaction` pane remains pending. Could this return an error instead and abort the process?
2. The polkadot.js app crashes after some time, which does not occur with `master` for me, without throwing an error in the shell output. It's happened at least 5 times during my testing so far, but I have not been able to reproduce deterministically, nor had time to debug.

**Comment by hoonsubin** [comment]
@Lederstrumpf thank you for the feedback! We'll work on it ASAP and let you know when the issues are fixed.

**Comment by mmagician** [comment]
@hoonsubin I hope all is fine on your side. Have you had any progress with the milestone recently?

**Comment by hoonsubin** [comment]
@mmagician Sorry for the silence.
Regarding the progress for the custom signature page, most of the issue regarding error handling with MetaMask is fixed. However, error handling with Ledger is still not fixed. There seems to be an issue with either MM itself, or the Ledger API (or Ledger bridge implementation in MM), as error callbacks (e.g., canceling signature requests from the device) on Ledger cannot be captured by the client app for unknown reasons, hence the infinite loading bug when the user cancels the signature from Ledger.
Error messages from MM are correctly captured, so I assume it has to do with rejections from Ledger being dropped by the MM provider. (Related line of code: https://github.com/MetaMask/eth-ledger-bridge-keyring/blob/main/index.js#L288)

I am still working on this issue, but due to other tasks, it may take longer than I expect. I will see what I can do before the end of this month. If I cannot find a simple solution, I'll take a hacky approach and add a signature request timeout function to manually throw an error.

**Comment by mmagician** [comment]
Thanks for the update and the explanation. That makes sense - from my point of view it's better to take the time and implement a proper, long-term solution. Good luck and keep us posted!

**Comment by semuelle** [comment]
Hi @hoonsubin, any news on this?

**Comment by hoonsubin** [comment]
> Hi @hoonsubin, any news on this?

@semuelle Sorry for the late response and the continued delay in the delivery. Our team was focusing on the Kusama Parachain auction for the past couple of months. Having said that, our auction has ended successfully now and we will be able to actively work on this task (finally)!
I will keep you posted and aim for delivering the task within this month, or early next month at the latest.

Thank you.

**Comment by hoonsubin** [comment]
@semuelle sorry for the long silence. I have noticed that the latest update of MetaMask fails to solve the issue I mentioned above. I have extended my investigation from last week, but this issue is definitely with the MetaMask Ledger iframe provider. So I have opened an issue on their repository hoping for some answers.
https://github.com/MetaMask/metamask-extension/issues/11680

Again, I apologize once again for the delay, but if this problem is solved within MetaMask, the application should automatically be fixed as well. I will stay in touch with MetaMask and Ledger bridge provider project and update my pull request on polkadot-js/apps and send a final review request when I can confirm that it works!

**Comment by semuelle** [comment]
Thanks for the update. No rush from our side. Keep us posted.

**Comment by alxs** [comment]
@hoonsubin thank you for your dedication. If you're done with the Trezor part, we can also evaluate milestone 1 separately and compensate you for it independently of the issues above. Feel free to move its delivery file to a separate pull request and we'll look into it.

**Comment by alxs** [comment]
@hoonsubin sorry we didn't get back to you on this. Unfortunately, I don't have access to a Ledger with which to test this. You'll probably have to finish the implementation without our feedback, but we'll double-check everything and test it on a device when the delivery is complete. Speaking of which, this delivery has now been open for over a year now. What's the status of https://github.com/polkadot-js/apps/pull/5903? Do you think you could wrap things up and complete the deliverables in the coming weeks, so we can take a final look at them and accept the milestones?

**Comment by hoonsubin** [comment]
@alxs Yes, I can confirm that the functionality is working well and fixes the issues you mentioned before. But due to some mistakes I made with Git, the commit history looks very dirty 😓 
I can can wrap things up and make it ready for review this week. Hopefully, we can finally finish this.

**Comment by hoonsubin** [comment]
Hello @alxs, https://github.com/polkadot-js/apps/pull/5903 is ready for review and merge

**Comment by alxs** [comment]
Thanks for the update @hoonsubin. I think @Lederstrumpf is best qualified to look at this delivery and has access to the devices on which to test it, he'll get back to you as soon as possible.

**Comment by alxs** [comment]
@hoonsubin I'm again really sorry for the wait. I will look into this delivery myself since unfortunately @Lederstrumpf no longer works with us. If you want to merge master into your branch again, I'll be able to review the PR in the same state it will be merged.

**Comment by hoonsubin** [comment]
> @hoonsubin I'm again really sorry for the wait. I will look into this delivery myself since unfortunately @Lederstrumpf no longer works with us. If you want to merge master into your branch again, I'll be able to review the PR in the same state it will be merged.

Thank you. I've updated my working branch https://github.com/polkadot-js/apps/pull/5903/commits/dbd685ee74a688aa1c3d0030a7fb4daea2a1a399

File: deliveries/polkadotjs-ecdsa-hardware-milestone-1.md
Comment (alxs) [inline deliveries/polkadotjs-ecdsa-hardware-milestone-1.md:13]: Could you update the documentation links in this and the milestone 2 files? Also, in your PR you link to https://docs.astar.network/build/metamask-signatures which doesn't work any longer. Lastly, could you also provide a link to the GitHub repo on which the documentation is hosted so that we can fork it?

File: deliveries/polkadotjs-ecdsa-hardware-milestone-1.md
Comment (alxs) [inline deliveries/polkadotjs-ecdsa-hardware-milestone-1.md:14]: This link doesn't seem to work either

File: deliveries/polkadotjs-ecdsa-hardware-milestone-2.md
Comment (alxs) [inline deliveries/polkadotjs-ecdsa-hardware-milestone-2.md:13]: Also here

**Comment by alxs** [comment]
By the way, pardon my ignorance but why is a custom pallet required for this in the first place?

File: deliveries/polkadotjs-ecdsa-hardware-milestone-1.md
Comment (alxs) [inline deliveries/polkadotjs-ecdsa-hardware-milestone-1.md:12]: Also, could you please update the commit hash here since this one is quite old and doesn't compile due to broken dependencies? I'm sorry, the problem is of course that we took so long in getting back to you.

File: deliveries/polkadotjs-ecdsa-hardware-milestone-2.md
Comment (alxs) [inline deliveries/polkadotjs-ecdsa-hardware-milestone-2.md:12]: And here

**Review by alxs — CHANGES_REQUESTED** [review]

File: deliveries/polkadotjs-ecdsa-hardware-milestone-1.md
Comment (hoonsubin) [reply deliveries/polkadotjs-ecdsa-hardware-milestone-1.md:13]: Sorry for the late response.
@akru can you update the file?

File: deliveries/polkadotjs-ecdsa-hardware-milestone-1.md
Comment (hoonsubin) [reply deliveries/polkadotjs-ecdsa-hardware-milestone-1.md:12]: @akru

**Comment by alxs** [comment]
@akru ping

**Comment by alxs** [comment]
@hoonsubin @akru are you two still planning to update this?

**Comment by alxs** [comment]
Please see my [comment](https://github.com/w3f/Grants-Program/pull/39#issuecomment-1158734893) on the application PR.

**Comment by hoonsubin** [comment]
> Please see my [comment](https://github.com/w3f/Grants-Program/pull/39#issuecomment-1158734893) on the application PR.

Sorry for the delay, seems like some documentation was lost in time. We will work on it again and get back to you ASAP

**Comment by akru** [comment]
Hello @alxs, documentation updated. Please check it again, I'm apologies for delay.

**Comment by alxs** [comment]
Thanks. I wanted to test it today but the instructions don't cover the setup, i.e. running a runtime with the pallet and a version of polkadot.js with the custom signatures page. Could you add some for this and also make sure the current steps can be followed? They seem to be for an older version of MetaMask. Also feel free to merge changes from upstream again, since your PR wasn't merged yet.

**Comment by alxs** [comment]
@hoonsubin ping

**Comment by hoonsubin** [comment]
> @hoonsubin ping

Sorry for the late response. We updated the document but found a breaking change on the pallet that was not reflected on the UI side. I will fix this week and get back to you ASAP!

**Comment by hoonsubin** [comment]
@alxs Here is the finished docs https://astarnetwork.notion.site/Backup-MetaMask-Support-ed399d80c11e4202a056776c20c64370 and https://github.com/polkadot-js/apps/pull/5903 has been fixed to support the latest breaking changes. Please review this

**Comment by alxs** [comment]
Thanks @hoonsubin! I'm connected to my local Astar node but there's no "Custom Signature" tab under Accounts, could you try to reproduce with a freshly cloned repo? I'm on Ubuntu and I used the latest Astar binary and the `page/custom-signature` branch in your apps clone.

**Comment by hoonsubin** [comment]
> Thanks @hoonsubin! I'm connected to my local Astar node but there's no "Custom Signature" tab under Accounts, could you try to reproduce with a freshly cloned repo? I'm on Ubuntu and I used the latest Astar binary and the `page/custom-signature` branch in your apps fork.
> 
> Note: I was using a Trezor, but probably the tab should always be there?

[Screencast from 07-25-2022 06:49:02 PM.webm](https://user-images.githubusercontent.com/40356749/180832414-def1252e-3dde-4108-817b-2e64c7e9fa2f.webm)

That is odd, it seems to be working well on my machine? Please check the attached video.

**Comment by alxs** [comment]
Thanks for checking, I think what happened is that I clicked on the link in the documentation under "Access the following [URL](https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A9944#/explorer)", which actually links to the live version of polkadot.js, and didn't double-check.

I tried to submit a transaction using a Trezor and got the following error message:

image

Have you tested it with one and could it be something on your end? Or is MetaMask supposed to handle anything device-specific?

---

Another thing - personally I don't think the timeout by default is a good solution since there's no reason to fail if nothing happened, which is confusing for the user. But it's definitely a "lesser of two evils" situation. At least I would modify the error message so it's clear why it timed out e.g.:
> Signature request timed out after 30 seconds.

---

I also tried testing it with a Ledger but I'm having another issue. I connected another (non-ledger) account and now I can't seem to connect the Ledger one. This is what happens when I click on Connect below:

image

This looks like the MetaMask connection is set up in a way that doesn't allow one to switch accounts?

**Comment by hoonsubin** [comment]
> Thanks for checking, I think what happened is that I clicked on the link in the documentation under "Access the following [URL](https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A9944#/explorer)", which actually links to the live version of polkadot.js, and didn't double-check.
> 
> I tried to submit a transaction using a Trezor and got the following error message:
> 
> image
> 
> Have you tested it with one and could it be something on your end? Or is MetaMask supposed to handle anything device-specific?
> 
> Another thing - personally I don't think the timeout by default is a good solution since there's no reason to fail if nothing happened, which is confusing for the user. But it's definitely a "lesser of two evils" situation. At least I would modify the error message so it's clear why it timed out e.g.:
> 
> > Signature request timed out after 30 seconds.
> 
> I also tried testing it with a Ledger but I'm having another issue. I connected another (non-ledger) account and now I can't seem to connect the Ledger one. This is what happens when I click on Connect below:
> 
> image
> 
> This looks like the MetaMask connection is set up in a way that doesn't allow one to switch accounts?

Thanks for the notice. I've assigned @bobo-k2, our client developer, to work on this task starting from next week. We will get back to you next week.

**Comment by hoonsubin** [comment]
Hello @alxs, sorry for the late response.
Me and our engineers checked, but we couldn't find the issue, as that problem should have been fixed with the latest changes. Could you update the branch and try it again?

**Comment by alxs** [comment]
I pointed out two different issues, which one are you referring to?

**Comment by hoonsubin** [comment]
> I pointed out two different issues, which one are you referring to?

The invalid signature error

**Comment by alxs** [comment]
That did happen to me with the latest changes, did you test it with a Trezor?

**Comment by alxs** [comment]
ping @hoonsubin

**Comment by alxs** [comment]
@akru @hoonsubin are you guys still planning to finish this? Every time we wait for another month between reviews, something breaks. It would be nice if you could go over it once again and make sure it's bulletproof (for both devices) so we can finally wrap this up.

**Comment by hoonsubin** [comment]
> @akru @hoonsubin are you guys still planning to finish this? Every time we wait for another month between reviews, something breaks. It would be nice if you could go over it once again and make sure it's bulletproof (for both devices) so we can finally wrap this up.

Sorry for not getting back to you earlier. But I want to emphasize that the error mentioned above (invalid signature) is already fixed, and the scope of this milestone is about Ledger (https://github.com/w3f/Grant-Milestone-Delivery/pull/83/files#diff-5c219b7e4c25cdba0f417f10ea9cd0dbd8812ce5112c8cb378f8bbd753bcc0d5R12), not Trezor. So we cannot guarantee the functionality of that device.

I would appreciate it if you could test it again with a Ledger and let us know if the issue still exists. We tested within our team from three developers, and the issue is resolved from our check.

Thank you

**Comment by alxs** [comment]
@hoonsubin you're delivering both milestones together in this PR. If you want to cancel milestone 1, feel free to submit an [amendment](https://github.com/w3f/Grants-Program#changes-to-a-grant-after-approval) to remove it from the application. Given the time frame of this grant, I think it's better not to postpone it to a later date.

**Comment by hoonsubin** [comment]
> @hoonsubin you're delivering both milestones together in this PR. If you want to cancel milestone 1, feel free to submit an [amendment](https://github.com/w3f/Grants-Program#changes-to-a-grant-after-approval) to remove it from the application. Given the time frame of this grant, I think it's better not to postpone it to a later date.

Sorry, this was my misunderstanding then. We internally agreed to only optimize for Ledger first and skip Trezor, but I didn't properly check the milestones in this PR.

I will do a local test on my end and decide wether to focus on only Ledger or not.

**Comment by hoonsubin** [comment]
@alxs 

https://user-images.githubusercontent.com/40356749/188599457-6a45a298-96c6-4467-8b29-09e3fa5464f6.mp4

I tried using it with Trezor, and it's working without any issues. Are you sure you're using the latest commit?

**Comment by alxs** [comment]
Yes. I just tested it again. Sorry if this is obvious, but did you try submitting a transaction?

Also a reminder to look into the other issue I [pointed out](https://github.com/w3f/Grant-Milestone-Delivery/pull/83#issuecomment-1198154878).

**Comment by hoonsubin** [comment]
> Yes. I just tested it again. Sorry if this is obvious, but did you try submitting a transaction?
> 
> Also a reminder to look into the other issue I [pointed out](https://github.com/w3f/Grant-Milestone-Delivery/pull/83#issuecomment-1198154878).

Sorry, I forgot to include the tx part in the video.

Okay, I will get back to you about the other issue. It seems like we need someone to dedicate this task

**Comment by hoonsubin** [comment]
> Yes. I just tested it again. Sorry if this is obvious, but did you try submitting a transaction?
> 
> Also a reminder to look into the other issue I [pointed out](https://github.com/w3f/Grant-Milestone-Delivery/pull/83#issuecomment-1198154878).

Sorry, it seems like I leftout the transaction part... But I also see inconsistent issues too... Looks like it's worth giving it a deeper look with the team.

Also for the second part, I changed it https://github.com/hoonsubin/apps/commit/47f32d3b34aebc07dfe153c2f2659e6f0f4d85eb

**Comment by alxs** [comment]
Sounds good. Let me know when it's working.

Thanks; however, I meant the connection issue. I can't switch accounts once I've connected one I couldn't figure out a way to fix this without uninstalling MetaMask.

**Comment by niklabh** [comment]
I am able to submit a transaction successfully with Physical Trezor Model T device. Firmware version 2.5.1.
[screen-capture (8).webm](https://user-images.githubusercontent.com/874046/189323043-81c759eb-e207-4841-bbc7-599209d74f2b.webm)

**Comment by niklabh** [comment]
> Sounds good. Let me know when it's working.
> 
> Thanks; however, I meant the connection issue. I can't switch accounts once I've connected one I couldn't figure out a way to fix this without uninstalling MetaMask.

For switching account you just need to switch in metamask and click "Load account from metamask" button. No need to uninstall metamask.

**Comment by alxs** [comment]
> I am able to submit a transaction successfully with Physical Trezor Model T device.

That doesn't mean the issue isn't there, as @hoonsubin and I both ran into it. I'm using a Trezor One with firmware version 1.11.2.

> For switching account you just need to switch in metamask and click "Load account from metamask" button. No need to uninstall metamask.

As I said, that doesn't work. Please see my original message.

**Comment by hoonsubin** [comment]
> > I am able to submit a transaction successfully with Physical Trezor Model T device.
> 
> 
> 
> That doesn't mean the issue isn't there, as @hoonsubin and I both ran into it. I'm using a Trezor One with firmware version 1.11.2.

I cannot seem to recreate this issue sadly. Considering the success rate, I would consider it as user error

**Comment by alxs** [comment]
I haven't been able to submit a successful transaction yet with my device. It's not "user error".

> But I also see inconsistent issues too, perhaps it's due to the nonce...

What did you mean by this then?

I think the transaction I submitted should have been invalid - perhaps it was incorrectly reported as the signature being invalid.

**Comment by hoonsubin** [comment]
> I haven't been able to submit a successful transaction yet with my device. It's not "user error".
> 
> 
> 
> > But I also see inconsistent issues too, perhaps it's due to the nonce...
> 
> 
> 
> What did you mean by this then?
> 
> 
> 
> I think the transaction I submitted should have been invalid - perhaps it was incorrectly reported as the signature being invalid.

It's difficult to debug this issue as I cannot reproduce it, and there isn't much info to work with. Can you also try submitting using a Trezor emulator and tell me if it still returns an error? Make sure you're using the latest commit

**Comment by alxs** [comment]
I tried to install the emulator and it was a very large package and the installation failed. Could you try sending a transaction from an account with insufficient balance first, and see what the error returned is? That might have been the issue. If that wasn't the case, I can try again.

**Comment by hoonsubin** [comment]
> I tried to install the emulator and it was a very large package and the installation failed. Could you try sending a transaction from an account with insufficient balance first, and see what the error returned is? That might have been the issue. If that wasn't the case, I can try again.

Sorry for the delay.

I've confirmed with three of our engineers and their physical Trezor that the signature is working well both for insufficient and sufficient balance accounts. I cannot seem to reproduce the error too.

**Comment by alxs** [comment]
Hey @hoonsubin, thanks for double-checking. I confirm that transactions still fail for me with an invalid signature error using the emulator, as you can see below. I have a local Astar node on v4.21.0 and the apps page on the latest commit in the `page/custom-signature` branch both running on an Ubuntu VPS, and the emulator locally on macOs Monterey as well as Metamask v10.18.4.

https://user-images.githubusercontent.com/15819210/191771967-1750a0f1-b4c4-4eae-ad52-33f59b0b1f14.mov

**Comment by hoonsubin** [comment]
> Hey @hoonsubin, thanks for double-checking. I confirm that transactions still fail for me with an invalid signature error using the emulator, as you can see below. I have a local Astar node on v4.21.0 and the apps page on the latest commit in the `page/custom-signature` branch both running on an Ubuntu VPS, and the emulator locally on macOs Monterey as well as Metamask v10.18.4.
> 
>  Screen.Recording.2022-09-22.at.16.16.23.mov

Then can you try Shibuya test net instead? I can send you some test net tokens if you give me your address

**Comment by alxs** [comment]
That still fails. Normal transactions from a polkadot{.js} account work fine.

Fyi, I will be ooo for the next two weeks - if you need assistance from someone in the grants team during that time, please ping another [evaluator](https://github.com/w3f/Grants-Program#w3f-grants-evaluators) (Sebastian is currently ooo). However, please keep in mind that getting the environment set up on my machine took a considerable amount of time, so ideally we wouldn't have to duplicate the effort. At any rate, I understand that this is not easy to debug if you can't replicate the error, but please understand that we can't really assist you with hands-on debugging as we don't have the resources to do this for every delivery.

**Comment by hoonsubin** [comment]
Hello @keeganquigley. Can you review this delivery, please?

**Comment by alxs** [comment]
@hoonsubin I'm back at work. Happy to review the delivery again as soon as it's fixed, but it doesn't look like you made any changes since our last exchange.

**Comment by hoonsubin** [comment]
> @hoonsubin I'm back at work. Happy to review the delivery again as soon as it's fixed, but it doesn't look like you made any changes since our last exchange.

As I recall, we could not recreate the issue you were encountering. May I request a second opinion from another reviewer's device?

**Comment by alxs** [comment]
@hoonsubin we only have one Trezor in the office for testing. Even if it worked with another device, the success rate should be 100% for us to accept the milestone. Did you try replicating my set-up as I mentioned [here](https://github.com/w3f/Grant-Milestone-Delivery/pull/83#issuecomment-1256200796)? If you're not interested in fixing this, we can always cancel the Trezor milestone and settle for Ledger integration.

**Comment by alxs** [comment]
That makes sense, thank you for your understanding. The amendment process is described [here](https://github.com/w3f/Grants-Program#changes-to-a-grant-after-approval), but it boils down to [editing](https://github.com/w3f/Grants-Program/edit/master/applications/polkadotjs-hardware.md) the existing proposal and submitting a PR. Once it's been accepted, I'll test the Ledger integration one last time and we should be good to go.

**Comment by hoonsubin** [comment]
> That makes sense, thank you for your understanding. The amendment process is described [here](https://github.com/w3f/Grants-Program#changes-to-a-grant-after-approval), but it boils down to [editing](https://github.com/w3f/Grants-Program/edit/master/applications/polkadotjs-hardware.md) the existing proposal and submitting a PR. Once it's been accepted, I'll test the Ledger integration one last time and we should be good to go.

Opened a PR https://github.com/w3f/Grants-Program/pull/1225

File: deliveries/polkadotjs-ecdsa-hardware-milestone-1.md
Comment (alxs) [inline deliveries/polkadotjs-ecdsa-hardware-milestone-1.md:5]:
```suggestion
* **Application Document:** https://github.com/w3f/Grants-Program/blob/master/applications/polkadotjs-hardware.md
```

**Comment by alxs** [comment]
@hoonsubin we would appreciate it if you could commit to a timeline on this. If you can't look into it in the coming weeks, could you please submit another amendment with an estimated delivery date?

**Comment by gluneau** [comment]
Hello @alxs, I've been testing this custom signature PR and I'm currently getting `Invalid Transaction: Transaction has a bad signature`,  we acknowledged the problem and we are in the process of fixing this.

**Comment by gluneau** [comment]
Good day @alxs, i've refactored the polkadot-js apps to take Hoon's implementation of the custom signature.  You can see the [changes](https://github.com/polkadot-js/apps/compare/master...gluneau:apps:cs-nov) for reference.

I've tested it on [ledger](https://shiden.subscan.io/extrinsic/2754599-2) and [trezor](https://shiden.subscan.io/extrinsic/2760260-2) and works great. 

I think I've been able to simplify the testing scenario by using the Shiden parachain RPCs.
So when you're ready, you can test your [local web app](http://localhost:3000/?rpc=wss%3A%2F%2Fshiden.public.blastapi.io#/custom-signature) on a live chain.

You'll need to seed your account, to do so after the first signature, copy the address it gives you and send it to me.  (Mine is `ZECSjh5SZjHW9zGek3TzTnn7vCbxgncDFbGZVKSCi4arEyw` for reference) I'll send you SDN to test.

image

To run the test, you can checkout my `cs-nov` [branch](https://github.com/gluneau/apps/tree/cs-nov) and `yarn start` it.  
Also notice there are 18 zeros needed to make a whole number.

Have fun!

**Comment by alxs** [comment]
@gluneau thanks for looking into this. Sounds good, could you send me some SDN to `bMvcmyxGTBKwYvghXqEtCwr9Zpxib44ZPTmduQFiXgDer3G`? When you're done, please also update the delivery file to point to the correct implementation of the `custom-signatures` pallet and `apps` page.

**Comment by niklabh** [comment]
Sent 50 SDN https://shiden.subscan.io/extrinsic/2764904-2

**Comment by alxs** [comment]
Thank you @akru @niklabh @gluneau. With this, I'm happy to let you know that your milestone has been accepted! You can find my evaluation notes [here](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/polkadotjs-ecdsa-hardware_1_alxs.md). One last thing: could you please also add a (Apache 2.0) LICENSE file to the documentation repository?

Please submit your invoice using the current [form](https://forms.gle/LSRr7PCjBpEbKGh89) and let me know when you're done. Make sure to include the same Kusama address as in your application and the amount denominated in USD(T).

**Comment by hilarious23** [comment]
Hi, we need to update the address to receive the grant.
I heard the payment would be made with USDT-statemint.
In this case, we would like to use `146iwtFpePG4GqBfNpMTDRgqfwNjj248JpvDKLyFPM1KSsa`.
Thank you.

**Comment by hoonsubin** [comment]
> Hi, we need to update the address to receive the grant. I heard the payment would be made with USDT-statemint. In this case, we would like to use `146iwtFpePG4GqBfNpMTDRgqfwNjj248JpvDKLyFPM1KSsa`. Thank you.

I'll open a PR to reflect these changes.

**Comment by hoonsubin** [comment]
> Thank you @akru @niklabh @gluneau. With this, I'm happy to let you know that your milestone has been accepted! You can find my evaluation notes [here](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/polkadotjs-ecdsa-hardware_1_alxs.md). One last thing: could you please also add a (Apache 2.0) LICENSE file to the documentation repository?
> 
> Please submit your invoice using the current [form](https://forms.gle/LSRr7PCjBpEbKGh89) and let me know when you're done. Make sure to include the same Kusama address as in your application and the amount denominated in USD(T).

Filled in the form!

**Comment by alxs** [comment]
Thanks, could you please include the Kusama address above in the invoice and denominate it in USDT?

**Comment by hoonsubin** [comment]
> Thanks, could you please include the Kusama address above in the invoice and denominate it in USDT?

Updated the address on the grant https://github.com/w3f/Grants-Program/pull/1382. The form will be updated shortly.

**Pull request merged**
